src/main/java/net/minecraftforge/common/world/WorldGenValidator.java

```diff
@@ -0,0 +1,121 @@
+import net.minecraftforge.fml.loading.FMLEnvironment;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public class WorldGenValidator {
+    private static final Logger LOGGER = LogManager.getLogger();
+    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();
+
+    /**
+     * Will check all biomes for any unregistered or broken worldgen elements.
+     * Please always register your configured features and such.
+     * Failure to do so may cause unregistered worldgen elements to nuke registered ones from the same biome.
+     */
+    public static void validateWorldGen(RegistryAccess.RegistryHolder registryHolder) {
+        // Do not run in production in order to not slowdown users with the checks
+        if (FMLEnvironment.production) return;
+
+        // Small cache so we can skip already checked out worldgen elements to speedup validating.
+        Set<Object> checkedElements = new HashSet<>();
+
+        // Grab all registries we will need.
+        Registry<Biome> biomeRegistry = registryHolder.registryOrThrow(Registry.BIOME_REGISTRY);
+        Registry<ConfiguredFeature<?, ?>> configuredFeatureRegistry = registryHolder.registryOrThrow(Registry.CONFIGURED_FEATURE_REGISTRY);
+        Registry<ConfiguredWorldCarver<?>> configuredCarverRegistry = registryHolder.registryOrThrow(Registry.CONFIGURED_CARVER_REGISTRY);
+        Registry<ConfiguredStructureFeature<?, ?>> configuredStructureRegistry = registryHolder.registryOrThrow(Registry.CONFIGURED_STRUCTURE_FEATURE_REGISTRY);
+
+        // Checks every biome for unregistered or broken worldgen elements.
+        for (Biome biome : biomeRegistry) {
+            BiomeGenerationSettings biomeGenerationSettings = biome.getGenerationSettings();
+            ResourceLocation biomeName = biome.getRegistryName();
+
+            biomeGenerationSettings.features().forEach(generationStageFeatureList ->
+                    generationStageFeatureList.forEach(configuredFeatureSupplier -> {
+                        ConfiguredFeature<?, ?> feature = configuredFeatureSupplier.get();
+                        if (!checkedElements.contains(feature)) {
+                            validate(feature, configuredFeatureRegistry, "ConfiguredFeature", ConfiguredFeature.DIRECT_CODEC, biomeName);
+                            checkedElements.add(feature);
+                        }
+                    }));
+
+            for (GenerationStep.Carving carvingStage : GenerationStep.Carving.values()) {
+                biomeGenerationSettings.getCarvers(carvingStage).forEach(configuredWorldCarverSupplier -> {
+                    ConfiguredWorldCarver<?> carver = configuredWorldCarverSupplier.get();
+                    if (!checkedElements.contains(carver)) {
+                        validate(carver, configuredCarverRegistry, "ConfiguredCarver", ConfiguredWorldCarver.DIRECT_CODEC, biomeName);
```

Comment:
The `"ConfiguredCarver"` within the string literal would best match the name of the class, `ConfiguredWorldCarver`.